cassiopeia/tools/lineage_tracing_simulator.py

```diff
@@ -0,0 +1,78 @@
+import abc
+
+import numpy as np
+
+from .tree import Tree
+
+
+class LineageTracingSimulator(abc.ABC):
+    r"""
+    Abstract base class for all lineage tracing simulators.
+
+    A LineageTracingSimulator implements a method overlay_lineage_tracing_data
+    which overlays lineage tracing data (i.e. character vectors) on top of the
+    tree. These are stored as the node's state.
+    """
+
+    @abc.abstractmethod
+    def overlay_lineage_tracing_data(self, tree: Tree) -> None:
+        r"""
+        Annotates the tree's nodes with lineage tracing character vectors.
+        These are stored as the node's state. (Operates on the tree in-place.)
+
+        Args:
+            tree: The tree to overlay lineage tracing data on.
+        """
+
+
+class IIDExponentialLineageTracer(LineageTracingSimulator):
+    r"""
+    Characters evolve IID over the lineage, with the same given mutation rate.
+
+    Args:
+        mutation_rate: The mutation rate of each character (same for all).
+        num_characters: The number of characters.
+    """
+
+    def __init__(self, mutation_rate: float, num_characters: float):
+        self.mutation_rate = mutation_rate
+        self.num_characters = num_characters
+
+    def overlay_lineage_tracing_data(self, tree: Tree) -> None:
+        r"""
+        See base class.
+        """
+        num_characters = self.num_characters
+        mutation_rate = self.mutation_rate
+
+        def dfs(node: int, tree: Tree):
+            node_state = tree.get_state(node)
+            for child in tree.children(node):
+                # Compute the state of the child
+                child_state = ""
+                edge_length = tree.get_age(node) - tree.get_age(child)
+                # print(f"{node} -> {child}, length {edge_length}")
+                assert edge_length >= 0
+                for i in range(num_characters):
+                    # See what happens to character i
+                    if node_state[i] != "0":
+                        # The character has already mutated; there in nothing
+                        # to do
+                        child_state += node_state[i]
+                        continue
+                    else:
+                        # Determine if the character will mutate.
+                        mutates = (
+                            np.random.exponential(1.0 / mutation_rate)
+                            < edge_length
+                        )
+                        if mutates:
+                            child_state += "1"
+                        else:
+                            child_state += "0"
+                tree.set_state(child, child_state)
+                dfs(child, tree)
```

Comment:
A general comment - a depth-first traversal function already exists in networkx.

But, if you want to do this a special way, I think this belongs as a method for the Tree class, not as a nested function here. Thoughts?

Reply:
I should look into the function from networkx that you mention. What exactly does it do? Can you please point me to it?

We need to do two things in `overlay_lineage_tracing_data`: (1) iterate over all edges down the tree (the tree traversal) (2) do some custom computation on each edge as we visit it. If we were to any of this behavior to the tree class I guess is would be (1)? IMO (2) should rest in the `LineageTracingSimulator` since it is something that varies.

Reply:
Yeah, that's a good point. The action items at each internal node in the DF traversal will change so it's not necessarily a general method for the tree class.

However, having a DFS method in the class that just returns nodes in post-order might be worthwhile and could cut down on some duplicated code in the other classes that operate on the `CassiopeiaTree` object. I'll add this a todo in the class.